src/graphics/mesh.rs

```diff
@@ -656,6 +686,156 @@ impl Drawable for Mesh {
     }
 }
 
+/// Mesh that will be rendered with hardware instancing.
+/// Use this when you have a lot of similar geometry which does not move around often.
+#[derive(Debug)]
+pub struct MeshBatch {
+    mesh: Mesh,
+    instance_properties: Vec<InstanceProperties>,
+    instance_buffer: Option<gfx::handle::Buffer<gfx_device_gl::Resources, InstanceProperties>>,
+}
+
+impl MeshBatch {
+    /// Creates a new mesh batch and allocates an instance buffer for it
+    pub fn new(
+        ctx: &mut Context,
+        mesh: Mesh,
+        instance_params: &[DrawParam],
+    ) -> GameResult<MeshBatch> {
+        let (instance_properties, instance_buffer) =
+            MeshBatch::create_instance_buffer(ctx, instance_params)?;
+
+        Ok(MeshBatch {
+            mesh: mesh,
+            instance_properties: instance_properties,
+            instance_buffer: instance_buffer,
+        })
+    }
+
+    /// Reset existing instance params to provided data and reallocate existing buffers
+    pub fn reset_instance_params(
+        &mut self,
+        ctx: &mut Context,
+        instance_params: &[DrawParam],
+    ) -> GameResult {
+        let (instance_properties, instance_buffer) =
+            MeshBatch::create_instance_buffer(ctx, instance_params)?;
+
+        self.instance_properties = instance_properties;
+        self.instance_buffer = instance_buffer;
+
+        Ok(())
+    }
+
+    /// Updates instance parameters in a batch without reallocating existing buffers
+    pub fn update_instance_params(
```

Comment:
Might want to think about the API for `*_instance_params()` and one way or another make it the same as `SpriteBatch`, so that both batch types work exactly the same way.

Reply:
That's a good idea, I'll do that. Do you think it's worth adding something like a "DrawableBatch" trait and implement it for SpriteBatch and MeshBatch?

As a side note, I don't really like that SpriteBatch calls flush() internally on every draw call - it really tanks performance, especially in debug builds. Since you've mentioned a graphics engine rewrite, I'd like to ask if there's any chance of swapping the global instance buffer to local instance buffers for separate batches that are not invalidated every frame?

Reply:
I wouldn't consider it worth making a trait to combine the two *yet*, since that would probably result in extra trait-type fluff when you have to say "what kind of drawable can I insert into this batch?"

For the side note, yeah, check out https://github.com/ggez/ggez/blob/adc36f8b151a1be30c43271094f340f0d70f9fda/src/graphics/mod.rs#L189 and associated stuff in https://github.com/ggez/ggez/tree/ggraphics/ggraphics .  Each `DrawCall` has its own instance buffer, and the order in which they are drawn and switched between becomes very explicit.  Any ideas, suggestions or wishlists you have for that codebase are welcome, I'm taking a break from it a little to chew on other problems.

Reply:
Sounds good, I'll check it out.